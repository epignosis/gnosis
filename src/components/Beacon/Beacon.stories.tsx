import React from "react";
import type { Meta, StoryFn } from "@storybook/react";
import Button from "../Button/Button";
import { Beacon, type BeaconProps } from "./Beacon";

const meta: Meta<typeof Beacon> = {
  title: "Components/Beacon",
  component: Beacon,
  argTypes: {
    placement: {
      control: { type: "radio" },
      options: ["left", "right"],
    },
    offsetX: { control: { type: "number" } },
    offsetY: { control: { type: "number" } },
    children: { control: false },
    isBeaconVisible: {
      control: { type: "boolean" },
    },
  },
};
export default meta;

const Template: StoryFn<BeaconProps> = (args) => (
  <Beacon {...args}>
    <Button>Beacon Target</Button>
  </Beacon>
);

export const Default = Template.bind({});
Default.args = {
  placement: "right",
  offsetX: 0,
  offsetY: 0,
  isBeaconVisible: true,
};

export const LeftPlacementWithCustomOffset = Template.bind({});
LeftPlacementWithCustomOffset.args = {
  placement: "left",
  offsetX: 0.25,
  offsetY: 0.25,
};

export const HiddenBeacon = Template.bind({});
HiddenBeacon.args = {
  placement: "right",
  offsetX: 0,
  offsetY: 0,
  isBeaconVisible: false,
};
