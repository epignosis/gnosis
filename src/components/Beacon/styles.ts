import { css, keyframes, SerializedStyles, Theme } from "@emotion/react";

// Keyframes
const grow = keyframes`
  0% { transform: scale(.1); }
  100% { transform: scale(1); }
`;

const active = keyframes`
  0% {
    transform: scale(.1);
    opacity: 1;
  }
  70% {
    transform: scale(2.5);
    opacity: 0;
  }
  100% {
    opacity: 0;
  }
`;

export const beaconContainer = (
  { beacon }: Theme,
  placement: "left" | "right",
  offsetX: string,
  offsetY: string,
): SerializedStyles => css`
  .beacon-content {
    position: relative;
    display: inline-block;
  }

  .beacon {
    position: absolute;
    top: 0;
    height: 1rem;
    width: 1rem;
    border-radius: 50%;
    background: ${beacon.color};
    animation: ${grow} 0.4s 1 linear;
    pointer-events: none;
    ${placement === "left"
      ? `
        left: 0;
        right: auto;
        transform: translate(calc(-1 * ${offsetX}), calc(-1 * ${offsetY})) translate(-50%, -50%);
      `
      : `
        right: 0;
        left: auto;
        transform: translate(${offsetX}, calc(-1 * ${offsetY})) translate(50%, -50%);
      `}
  }

  .beacon::before {
    content: "";
    position: absolute;
    inset: 0;
    border-radius: 50%;
    box-shadow: 0 0 2px 2px ${beacon.color};
    animation: ${active} 2s infinite linear;
  }
`;
