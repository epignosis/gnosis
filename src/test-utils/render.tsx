import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
