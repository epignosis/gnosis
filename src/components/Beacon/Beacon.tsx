import React, { PropsWithChildren } from "react";
import { SerializedStyles } from "@emotion/react";
import { beaconContainer } from "./styles";

type Placement = "left" | "right";

export type BeaconProps = PropsWithChildren<{
  placement?: Placement;
  /** Optional fine-tuning if you need to nudge the beacon in X-axis (rem units). */
  offsetX?: number;
  /** Optional fine-tuning if you need to nudge the beacon in Y-axis (rem units). */
  offsetY?: number;
  isBeaconVisible?: boolean;
}>;

export const Beacon: React.FC<BeaconProps> = ({
  children,
  placement = "right",
  offsetX = 0,
  offsetY = 0,
  isBeaconVisible = true,
}) => {
  return (
    <div
      css={(theme): SerializedStyles =>
        beaconContainer(theme, placement, `${offsetX}rem`, `${offsetY}rem`)
      }
    >
      <span className="beacon-content">
        {children}
        {isBeaconVisible && <span className="beacon" />}
      </span>
    </div>
  );
};
