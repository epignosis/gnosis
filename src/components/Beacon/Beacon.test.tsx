import React from "react";
import { Beacon } from "./Beacon";
import { render, screen } from "@test-utils/render";

describe("<Beacon />", () => {
  it("renders children and beacon", () => {
    render(
      <Beacon>
        <span data-testid="child">Test Child</span>
      </Beacon>,
    );
    const child = screen.getByTestId("child");
    const beacon = document.querySelector(".beacon-content .beacon");
    expect(child).toHaveTextContent("Test Child");
    expect(beacon).toBeInTheDocument();
  });

  it("renders with left placement and custom offsets", () => {
    render(
      <Beacon placement="left" offsetX={0.25} offsetY={0.25}>
        <span>Left Child</span>
      </Beacon>,
    );
    const child = screen.getByText("Left Child");
    expect(child).toBeInTheDocument();
    // Style assertions can be added with jest-emotion or snapshot
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Beacon>
        <span>Snapshot Child</span>
      </Beacon>,
    );
    expect(container).toMatchSnapshot();
  });

  it("does not render beacon when isBeaconVisible is false", () => {
    render(
      <Beacon isBeaconVisible={false}>
        <span>No Beacon</span>
      </Beacon>,
    );
    const beacon = document.querySelector(".beacon-content .beacon");
    expect(beacon).not.toBeInTheDocument();
  });

  it("applies correct styles for right placement and default offset", () => {
    render(
      <Beacon placement="right" offsetX={0} offsetY={0}>
        <span>Styled Child</span>
      </Beacon>,
    );
    const beacon = document.querySelector(".beacon-content .beacon") as HTMLElement;
    const computed = window.getComputedStyle(beacon);
    //Beacon size
    expect(computed.height).toBe("1rem");
    expect(computed.width).toBe("1rem");

    //Beacon starting position
    expect(computed.right).toBe("0px");
    expect(computed.top).toBe("0px");

    // Check transform contains expected translate for right placement
    expect(computed.transform).toContain("translate(0rem, calc(-1 * 0rem)) translate(50%, -50%)");
  });

  it("applies correct styles for left placement and custom offset", () => {
    render(
      <Beacon placement="left" offsetX={0.25} offsetY={0.25}>
        <span>Styled Left Child</span>
      </Beacon>,
    );
    const beacon = document.querySelector(".beacon-content .beacon") as HTMLElement;
    const computed = window.getComputedStyle(beacon);
    //Beacon size
    expect(computed.height).toBe("1rem");
    expect(computed.width).toBe("1rem");

    //Beacon starting position
    expect(computed.left).toBe("0px");
    expect(computed.top).toBe("0px");

    // Check transform contains expected translate for left placement with custom offset
    expect(computed.transform).toContain(
      "translate(calc(-1 * 0.25rem), calc(-1 * 0.25rem)) translate(-50%, -50%)",
    );
  });
});
