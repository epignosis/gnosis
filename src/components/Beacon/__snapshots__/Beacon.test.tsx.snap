// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Beacon /> matches snapshot 1`] = `
<div>
  <div
    class="css-tt1vh5-beaconContainer-beaconContainer"
  >
    <span
      class="beacon-content"
    >
      <span>
        Snapshot Child
      </span>
      <span
        class="beacon"
      />
    </span>
  </div>
</div>
`;
